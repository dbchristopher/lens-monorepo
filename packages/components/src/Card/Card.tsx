import omit from 'lodash/omit';
import React, { FunctionComponent, Ref } from 'react';
import styled, { StyledComponent, withTheme } from 'styled-components';
import { ThemedProps } from '../types';
import { BoxProps } from '../Box/Box';
import Box from '../Box';

export interface CardProps extends BoxProps<HTMLDivElement> {
  raised?: boolean;
}

export type CardComponentType = FunctionComponent<ThemedProps<CardProps>>;
export type StyledCardComponentType = StyledComponent<
  CardComponentType,
  CardProps
>;

const InternalCard: CardComponentType = (props: ThemedProps<CardProps>) => (
  <Box
    bg="palette.white"
    borderRadius={props.theme.radii.medium}
    border={`solid 1px ${props.theme.colors.semanticColors.primary.borderColor}`}
    display="flex"
    flexDirection="column"
    height="100%"
    min-width="200px"
    overflow="hidden"
    {...omit(props, ['raised'])}
  >
    {props.children}
  </Box>
);

const InternalCardThemed = withTheme(InternalCard);

const CardFactory = React.forwardRef<StyledCardComponentType, CardProps>(
  (props: CardProps, ref: Ref<StyledCardComponentType>) => (
    <InternalCardThemed ref={ref} {...props} />
  )
);

const Card: StyledCardComponentType = styled<CardComponentType>(CardFactory)`
  box-shadow: ${props => (props.raised ? props.theme.shadows[1] : 'none')};
  transition: border ${props => props.theme.transitions.durationQuick}
      ${props => props.theme.easings.ease},
    box-shadow ${props => props.theme.transitions.durationQuick}
      ${props => props.theme.easings.ease};
  &:hover {
    border-color: ${props => props.theme.colors.palette.charcoal300};
    box-shadow: ${props => (props.raised ? props.theme.shadows[2] : 'none')};
  }
`;

export default Card;
