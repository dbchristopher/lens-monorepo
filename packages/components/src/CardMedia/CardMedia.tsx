import React, { FunctionComponent, Ref } from 'react';
import styled, { StyledComponent } from 'styled-components';
import { backgroundPosition, BackgroundPositionProps } from 'styled-system';
import { BoxProps } from '../Box/Box';
import Box from '../Box';

export interface CardMediaProps
  extends BoxProps<HTMLDivElement>,
    BackgroundPositionProps {
  image: string;
}
type ComponentType = FunctionComponent<CardMediaProps>;
type StyledComponentType = StyledComponent<ComponentType, CardMediaProps>;

const InternalCardMedia: React.FC<CardMediaProps> = ({ image, ...props }) => (
  <Box {...props}>{props.children}</Box>
);

const CardMediaFactory = React.forwardRef<StyledComponentType, CardMediaProps>(
  (props: CardMediaProps, ref: Ref<StyledComponentType>) => (
    <InternalCardMedia ref={ref} {...props} />
  )
);

const CardMedia = styled<ComponentType>(CardMediaFactory)`
  background-size: cover;
  background-repeat: no-repeat;
  ${backgroundPosition}
  background-image: url(${props => props.image});
  overflow: hidden;
  height: 0;
  padding-top: 56%;
`;

export default CardMedia;
