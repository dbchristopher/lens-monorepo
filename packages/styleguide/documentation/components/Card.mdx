```js noeditor
const StatusAndResources = require('../../../styleguide_components/StatusAndResources')
  .StatusAndResources
;<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/PUkKDfseVgoPGNJng7TJ2TIJ/Card"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Card/Card.tsx"
  feedbackTitle="Card Component Feedback"
/>
```

<div class="doc-section-divider"></div>

### A basic Card

At its most basic, a `Card` is a slightly styled container that organizes groups of content, usually linking to a more in-depth view of that content. It can have elevation and a border. A default `Card` makes no assumptions about the content you place inside of it or the size of your card, but you should follow the guidelines to be sure you use them effectively.

```js
// A Card
// Try typing some content into the Card to see how it handles content by default
import { Card } from './Card'
import { Text } from '../Text'
;<Card raised>
  <Text mode="subdued" fontSize="xlarge">
    Hello World!
  </Text>
</Card>
```

<div class="doc-section-divider"></div>

### A Card with content

Using the [`CardContent`](/#!/CardContent) component quickly gives your content consistent spacing inside your `Card`. You can then use other components to layout your `Card`.

```js
import { Card, CardContent } from './'
import { Heading } from '../Heading'
import { Text } from '../Text'
;<Card raised>
  <CardContent>
    <Heading fontSize="xxxlarge">ðŸŽ‰ Looker Release Notes ðŸŽ‰</Heading>
    <Heading is="h4" fontSize="small">
      Read all about our latest features
    </Heading>
    <Text fontSize="xsmall" mode="subdued">
      Last updated 3 days ago
    </Text>
  </CardContent>
</Card>
```

<div class="doc-section-divider"></div>

### A group of media Cards

A few common patterns for `Cards` are displaying them in groups and using images to reinforce the content of the card.

```js
import { Card } from './Card'
import { CardContent } from './CardContent'
import { CardMedia } from './CardMedia'
import { Flex } from '../Flex'
import { FlexItem } from '../FlexItem'
import { Heading } from '../Heading'
import { Text } from '../Text'
;<Flex>
  <FlexItem width="33%" mr="small">
    <Card raised>
      <CardMedia
        image="https://placeimg.com/640/480/nature"
        title="Summer Nature"
      />
      <CardContent>
        <Text
          fontSize="xsmall"
          textTransform="upper"
          fontWeight="semiBold"
          mode="subdued"
        >
          Summer
        </Text>
        <Heading is="h4" fontSize="medium" fontWeight="semiBold" truncate>
          Life in The Great Outdoors
        </Heading>
        <div>
          <Text fontSize="small">
            10 reasons to get off the couch and head outside this summer.
          </Text>
        </div>
      </CardContent>
    </Card>
  </FlexItem>
  <FlexItem width="33%" mr="small">
    <Card raised>
      <CardMedia
        image="https://placeimg.com/630/480/nature"
        title="A Scenic Valley"
      />
      <CardContent>
        <Text
          fontSize="xsmall"
          textTransform="upper"
          fontWeight="semiBold"
          mode="subdued"
        >
          Explore
        </Text>
        <Heading is="h4" fontSize="medium" fontWeight="semiBold" truncate>
          Best Scenic Hikes
        </Heading>
        <div mt="xs">
          <Text fontSize="small">
            Looking for a new place to trailblaze? Make sure it has a great
            view!
          </Text>
        </div>
      </CardContent>
    </Card>
  </FlexItem>
  <FlexItem width="33%">
    <Card raised>
      <CardMedia
        image="https://placeimg.com/620/480/nature"
        title="Relaxing Views"
      />
      <CardContent>
        <Text
          fontSize="xsmall"
          textTransform="upper"
          fontWeight="semiBold"
          mode="subdued"
        >
          Relax
        </Text>
        <Heading is="h4" fontSize="large" fontWeight="semiBold" truncate>
          Mindfull Wilderness
        </Heading>
        <div mt="xs">
          <Text fontSize="small">Find a place to find your self.</Text>
        </div>
      </CardContent>
    </Card>
  </FlexItem>
</Flex>
```
