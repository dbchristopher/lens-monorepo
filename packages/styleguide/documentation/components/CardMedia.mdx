```js noeditor
const StatusAndResources = require('../../../styleguide_components/StatusAndResources')
  .StatusAndResources

;<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/PUkKDfseVgoPGNJng7TJ2TIJ/Card"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Card/CardMedia.tsx"
  feedbackTitle="Card Media Component Feedback"
/>
```

### Displaying an image in your card

A common pattern for `Cards` is to display an image that reinforces what the content is about.

The `CardMedia` component accepts and `image` property that will render the image as a background image and the optional `title` property used to give the image an accessible title if needed.

```js
import { Card } from './Card'
import { CardContent } from './CardContent'
import { CardMedia } from './CardMedia'
import { Box } from '../Box'
import { Heading } from '../Heading'
import { Text } from '../Text'
;<>
  <Box>
    <Card>
      <CardMedia image="https://www.fillmurray.com/300/150" title="Bill 1" />
      <CardContent>
        <Heading>A wild Bill appears</Heading>
      </CardContent>
    </Card>
  </Box>
  <Box mt="medium">
    <Card>
      <CardMedia image="https://www.fillmurray.com/600/300" title="Bill 2" />
      <CardContent>
        <Heading>Another Bill spoting</Heading>
      </CardContent>
    </Card>
  </Box>
</>
```
