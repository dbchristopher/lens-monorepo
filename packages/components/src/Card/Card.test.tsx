import 'jest-styled-components'
import * as React from 'react'
import { createWithTheme } from '../../../test/utils/create_with_theme'
import { Card } from './Card'
import { CardMedia } from './CardMedia'

test('A Card', () => {
  const component = createWithTheme(<Card>ðŸ¥‘</Card>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A raised Card', () => {
  const component = createWithTheme(<Card raised={true}>ðŸ¥‘</Card>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A raised Card with a CardMedia block', () => {
  const component = createWithTheme(
    <Card raised={true}>
      <CardMedia image="http://placekitten.com/200/300" />
    </Card>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
