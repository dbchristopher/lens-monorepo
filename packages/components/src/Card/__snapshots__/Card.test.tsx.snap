// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Card 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  border: solid 1px #C1C6CC;
  border-radius: 0.25rem;
  background-color: #FFFFFF;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
  overflow: hidden;
}

.c0 {
  box-shadow: none;
  -webkit-transition: border 150ms cubic-bezier(0.86,0,0.07,1), box-shadow 150ms cubic-bezier(0.86,0,0.07,1);
  transition: border 150ms cubic-bezier(0.86,0,0.07,1), box-shadow 150ms cubic-bezier(0.86,0,0.07,1);
}

.c0:hover {
  border-color: #C1C6CC;
  box-shadow: none;
}

<div
  className="c0 c1"
  min-width="200px"
>
  ðŸ¥‘
</div>
`;

exports[`A raised Card 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  border: solid 1px #C1C6CC;
  border-radius: 0.25rem;
  background-color: #FFFFFF;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
  overflow: hidden;
}

.c0 {
  box-shadow: 0px 1px 8px rgba(0,0,0,0.08),0px 1px 1px rgba(0,0,0,0.05);
  -webkit-transition: border 150ms cubic-bezier(0.86,0,0.07,1), box-shadow 150ms cubic-bezier(0.86,0,0.07,1);
  transition: border 150ms cubic-bezier(0.86,0,0.07,1), box-shadow 150ms cubic-bezier(0.86,0,0.07,1);
}

.c0:hover {
  border-color: #C1C6CC;
  box-shadow: 0 2px 12px rgba(0,0,0,0.11),0 1px 4px rgba(0,0,0,0.04);
}

<div
  className="c0 c1"
  min-width="200px"
>
  ðŸ¥‘
</div>
`;

exports[`A raised Card with a CardMedia block 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  border: solid 1px #C1C6CC;
  border-radius: 0.25rem;
  background-color: #FFFFFF;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
  overflow: hidden;
}

.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
}

.c0 {
  box-shadow: 0px 1px 8px rgba(0,0,0,0.08),0px 1px 1px rgba(0,0,0,0.05);
  -webkit-transition: border 150ms cubic-bezier(0.86,0,0.07,1), box-shadow 150ms cubic-bezier(0.86,0,0.07,1);
  transition: border 150ms cubic-bezier(0.86,0,0.07,1), box-shadow 150ms cubic-bezier(0.86,0,0.07,1);
}

.c0:hover {
  border-color: #C1C6CC;
  box-shadow: 0 2px 12px rgba(0,0,0,0.11),0 1px 4px rgba(0,0,0,0.04);
}

.c2 {
  background-size: cover;
  background-repeat: no-repeat;
  background-image: url(http://placekitten.com/200/300);
  overflow: hidden;
  height: 0;
  padding-top: 56%;
}

<div
  className="c0 c1"
  min-width="200px"
>
  <div
    className="c2 c3"
  />
</div>
`;
